packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js

```diff
@@ -86,37 +132,128 @@ describe('persist-upload', () => {
+        sandbox.restore();
+    });
+
+    it('should return null if agencyId is undefined', async () => {
+        const agencyId = undefined;
+        const validAgencyId = await getValidAgencyId(agencyId);
+        expect(validAgencyId).to.equal(null);
     });
 
+    it('should throw a ValidationError if agencyId is not associated with current tenant', async () => {
+        // In this case, the user is in tenant 0, which has agency list [Agency1, Agency2], and the agencyId is 3, which is not in the list
+        const userId = 1;
+        const agencyId = 3;
+
+        try {
+            // eslint-disable-next-line no-unused-vars
+            const validAgencyId = await getValidAgencyId(agencyId, userId);
+            throw new Error('Expected an error to be thrown');
+        } catch (err) {
+            expect(err).to.be.an.instanceof(ValidationError);
+            expect(err.message).to.equal('Supplied agency ID 3 does not correspond to an agency in the user\'s tenant 0');
+        }
+    });
+
+    it('should return the agencyId if it corresponds to the tenant agency list', async () => {
+        // In this case, the user is in tenant 0, which has agency list [Agency1, Agency2], and the agencyId is 2, which is in the list
+        const userId = 1;
+        const agencyId = 2;
+
+        const validAgencyId = await getValidAgencyId(agencyId, userId);
+        expect(validAgencyId).to.equal(agencyId);
+    });
+});
+
+describe('getValidReportingPeriodId', () => {
+    const getValidReportingPeriodId = persistUploadModule.__get__('getValidReportingPeriodId');
+    const getReportingPeriod = sinon.stub();
+    persistUploadModule.__set__('getReportingPeriod', getReportingPeriod);
+    let sandbox;
+
+    beforeEach(() => {
+        sandbox = sinon.createSandbox();
+        getReportingPeriod.resetHistory();
+    });
+
+    afterEach(() => {
+        sandbox.restore();
+    });
+
+    it('should fall back to the current reporting period if reportingPeriodId is undefined', async () => {
+        const reportingPeriodId = undefined;
+        const getReportingPeriodStub = sandbox.stub().returns({ id: 1 });
+        persistUploadModule.__set__('getReportingPeriod', getReportingPeriodStub);
+        const validReportingPeriodId = await getValidReportingPeriodId(reportingPeriodId);
+        expect(validReportingPeriodId).to.equal(1);
+    });
+
+    it('should return the ID of the reporting period when it exists', async () => {
+        const reportingPeriod = { id: 123 };
+        getReportingPeriod.resolves(reportingPeriod);
+        persistUploadModule.__set__('getReportingPeriod', getReportingPeriod);
+
+        const result = await getValidReportingPeriodId(reportingPeriod.id);
+
+        expect(getReportingPeriod.calledOnceWith(reportingPeriod.id)).to.be.true;
+        expect(result).to.equal(reportingPeriod.id);
+    });
+
+    it('should throw a ValidationError when the reporting period does not exist', async () => {
```

Comment:
In addition to checking the "does not exist" case, a more critical safety check might be the "reporting period does exist for a different tenant" case

Reply:
Interesting! I was initially thinking this was covered by the `reporting-period.spec.js` test, and thus a few layers deeper than the "unit" of this method, but then I realized, `getReportingPeriod` actually returns `undefined` if the conditions aren't met, so we can represent this with the mocks - I added a test to represent this